AX: aria-hidden=false should be a synonym of undefined

hoffmanjoshua opened the pull request:
#### b365574687fc38236c3abe1a5c1cc8e3fef0f712
<pre>
AX: aria-hidden=false should be a synonym of undefined
<a href="https://bugs.webkit.org/show_bug.cgi?id=267150">https://bugs.webkit.org/show_bug.cgi?id=267150</a>
<a href="https://rdar.apple.com/120557669">rdar://120557669</a>

Reviewed by Tyler Wilcock.

Per spec changes (see <a href="https://github.com/w3c/aria/pull/2090)">https://github.com/w3c/aria/pull/2090)</a>, we should treat `aria-hidden=false` as undefined.

This patch removes support for aria-hidden=false, updating places where we relied on the behaviors of isNodeARIAVisible
with the new behavior. For example, we need to check if a child node is focused before deciding whether to skip it if it is
aria-hidden (tested by accessibility/datetime/input-date-field-labels-and-value-changes.html).

Tests that explicitly validate aria-hidden false were removed, and a new test to check that we are ignoring this property
has been added.

* LayoutTests/accessibility/aria-hidden-false-ignored-expected.txt: Added.
* LayoutTests/accessibility/aria-hidden-false-ignored.html: Added.
* LayoutTests/accessibility/aria-hidden-false-works-in-subtrees.html: Removed.
* LayoutTests/accessibility/aria-hidden-negates-no-visibility.html: Removed.
* LayoutTests/accessibility/aria-hidden-subtree-expected.txt: Added.
* LayoutTests/accessibility/aria-hidden-subtree.html: Added.
* LayoutTests/accessibility/aria-modal-expected.txt:
* LayoutTests/accessibility/aria-modal.html:
* LayoutTests/accessibility/aria-visible-element-roles.html: Removed.
* LayoutTests/accessibility/datetime/input-date-field-labels-and-value-changes.html:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::modalElementHasAccessibleContent):
(WebCore::AXObjectCache::isNodeVisible const):
(WebCore::AXObjectCache::getOrCreate):
(WebCore::isNodeFocused):
(WebCore::isNodeAriaVisible): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::textUnderElement const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::defaultObjectInclusion const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::addNodeOnlyChildren):

Canonical link: <a href="https://commits.webkit.org/284905@main">https://commits.webkit.org/284905@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53b1903247475d06a4bf4abd87c56a44cb2d20c2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70852 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50262 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23621 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74952 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22053 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72968 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58060 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21873 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56077 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14552 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73918 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45687 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61086 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36523 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42342 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18520 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20394 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64277 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18881 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76666 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15082 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18081 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15126 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61149 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63762 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15692 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11839 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5486 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46063 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/834 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47135 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48416 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46877 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -4658,44 +4660,12 @@ void AXObjectCache::deferTextReplacementNotificationForTextControl(HTMLTextFormC
 {
-    // If an element is focused, it should not be hidden.
     if (RefPtr element = dynamicDowncast<Element>(node); element && element->focused())
         return true;
```
Could this be:

```
return is<Element>(node) && uncheckedDowncast<Element>(node)->focused();
```

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -882,7 +884,7 @@ AccessibilityObject* AXObjectCache::getOrCreate(Node& node, IsPartOfRelation isP
         bool hasDisplayContents = element && element->hasDisplayContents();
         bool isPopover = element && element->hasAttributeWithoutSynchronization(popoverAttr);
-        if (!inCanvasSubtree && !insideMeterElement && !hasDisplayContents && !isPopover && !isNodeAriaVisible(node))
+        if (!inCanvasSubtree && !insideMeterElement && !hasDisplayContents && !isPopover && !isNodeFocused(node))
```
Do we need to use `isNodeFocused()` here? We already do the work to cast to `Element`. What about:

```
&& (!element || !element->focused())`
```

hoffmanjoshua replied:
Yeah this should work, we might be able to get rid of `isNodeFocused` altogether.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -446,8 +446,10 @@ bool AXObjectCache::isNodeVisible(Node* node) const
-    if (!isNodeAriaVisible(*node))
-        return false;
+    if (auto* element = dynamicDowncast<Element>(node)) {
+        if (equalLettersIgnoringASCIICase(element->attributeWithDefaultARIA(aria_hiddenAttr), "true"_s) && !isNodeFocused(*node))
```
Can we use `!element->focused()` instead here?

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -446,8 +446,10 @@ bool AXObjectCache::isNodeVisible(Node* node) const
     // We also need to consider aria hidden status.
-    if (!isNodeAriaVisible(*node))
-        return false;
+    if (auto* element = dynamicDowncast<Element>(node)) {
```
We already do an `is<Element>(node)` at the top of this function, so this is redundant. I propose we add this to the top of the function, and avoid the need for a cast on this line completely:

```cpp
RefPtr element = dynamicDowncast<Element>(node);
if (!element)
    return false;
```

hoffmanjoshua replied:
I like that, definitely cleaner. Will make that change!

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -446,7 +447,7 @@ bool AXObjectCache::isNodeVisible(Node* node) const
+    if (equalLettersIgnoringASCIICase(element->attributeWithDefaultARIA(aria_hiddenAttr), "true"_s) && !element->focused())
         return false;
 
     return true;
```
I think you can write more concisely as:
return !equalLettersIgnoringASCIICase(element->attributeWithDefaultARIA(aria_hiddenAttr), "true"_s) || element->focused();

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -4658,44 +4659,9 @@ void AXObjectCache::deferTextReplacementNotificationForTextControl(HTMLTextFormC
-    }
-    
-    return !requiresAriaHiddenFalse || ariaHiddenFalsePresent;
+    return is<Element>(node) && uncheckedDowncast<Element>(node).focused();
```
Can we remove the aria-hidden=true check in the walk up?

hoffmanjoshua replied:
I think I'll need to add that back in (probably in `isARIAHidden()`). Doesn't look like we have a test covering that, so I'll add that in to my new test as well.

hoffmanjoshua replied:
I addressed this in the latest revision by using `isAXHidden()` (with the appropriate inverse logic) in places where we used to do `isNodeARIAVisible()`. `isAXHidden()` takes care of the ancestor walk-up to check for ARIA hidden.

When computing `defaultObjectInclusion()`, however, we already perform an ancestor walk to check for `aria-hidden=true`, so that case is covered already:
```    
    if (isARIAHidden())
        return AccessibilityObjectInclusion::IgnoreObject;

    bool ignoreARIAHidden = isFocused();
    if (Accessibility::findAncestor<AccessibilityObject>(*this, false, [&] (const auto& object) {
        return (!ignoreARIAHidden && object.isARIAHidden()) || object.ariaRoleHasPresentationalChildren() || !object.canHaveChildren();
    }))
        return AccessibilityObjectInclusion::IgnoreObject;
 ```

hoffmanjoshua commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -351,7 +351,8 @@ bool AXObjectCache::modalElementHasAccessibleContent(Element& element)
                 // Its content is exposed by its parent.
                 // Treat such elements as having accessible content.
-                if (axObject->roleValue() == AccessibilityRole::StaticText)
+                // FIXME: This may not be sufficient for visibility:hidden or inert (https://bugs.webkit.org/show_bug.cgi?id=280914)
```
@twilco here is the updated comment + condition (with new bug https://bugs.webkit.org/show_bug.cgi?id=280914 filed)

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -882,7 +881,7 @@ AccessibilityObject* AXObjectCache::getOrCreate(Node& node, IsPartOfRelation isP
         bool hasDisplayContents = element && element->hasDisplayContents();
         bool isPopover = element && element->hasAttributeWithoutSynchronization(popoverAttr);
-        if (!inCanvasSubtree && !insideMeterElement && !hasDisplayContents && !isPopover && !isNodeAriaVisible(node))
+        if (!inCanvasSubtree && !insideMeterElement && !hasDisplayContents && !isPopover && (!element || !element->focused()))
```
Could we use `isNodeFocused` here? Might read a bit better.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2317,22 +2317,16 @@ static bool shouldPrependSpace(AXCoreObject& object, AXCoreObject* previousObjec
     RefPtr node = this->node();
     if (auto* text = dynamicDowncast<Text>(node.get()))
-        return !mode.isHidden() || isAriaVisible() ? text->wholeText() : emptyString();
+        return !mode.isHidden() && !isAXHidden() ? text->wholeText() : emptyString();
```
I built this patch and removed the `&& !isAXHidden()` check you added here, and no tests failed. Do we actually need this?

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2317,22 +2317,16 @@ static bool shouldPrependSpace(AXCoreObject& object, AXCoreObject* previousObjec
     // referenced by aria-labelledby or aria-describedby, and is not a host language text
     // alternative, the empty string should be returned.
-    if (mode.isHidden() && !isAriaVisible() && (node && !ancestorsOfType<HTMLCanvasElement>(*node).first())) {
+    if ((mode.isHidden() || isAXHidden()) && (node && !ancestorsOfType<HTMLCanvasElement>(*node).first())) {
```
I built this patch and removed the`|| isAXHidden()` check you added here, and no tests failed. Do we actually need this?

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2456,7 +2456,7 @@ void AccessibilityRenderObject::addNodeOnlyChildren()
         }
 
-        if (!nodeHasDisplayContents(*child) && !isNodeAriaVisible(*child))
+        if (!nodeHasDisplayContents(*child) || isAXHidden())
```
Based on our discussion in-person, I think we don't need this `|| isAXHidden()` check?

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -4049,10 +4049,6 @@ AccessibilityObjectInclusion AccessibilityObject::defaultObjectInclusion() const
-                return AccessibilityObjectInclusion::DefaultBehavior;
-
             return AccessibilityObjectInclusion::IgnoreObject;
         }
```
The curly-braces associated with this if-statement are unnecessary now.

hoffmanjoshua replied:
Thanks for catching that—surprising that the styler checker didn't call that out

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2314,22 +2314,16 @@ static bool shouldPrependSpace(AXCoreObject& object, AXCoreObject* previousObjec
     // referenced by aria-labelledby or aria-describedby, and is not a host language text
     // alternative, the empty string should be returned.
-    if (mode.isHidden() && !isAriaVisible() && (node && !ancestorsOfType<HTMLCanvasElement>(*node).first())) {
+    if ((mode.isHidden() && (node && !ancestorsOfType<HTMLCanvasElement>(*node).first())) {
```
Is this added parentheses to the left of `mode.isHidden()` necessary? I'd think the following would suffice, but maybe I'm missing something 🤔 

`if (mode.isHidden() && node && !ancestorsOfType<HTMLCanvasElement>(*node).first())`

hoffmanjoshua replied:
Yeah we can get rid of that

hoffmanjoshua committed:
AX: aria-hidden=false should be a synonym of undefined
https://bugs.webkit.org/show_bug.cgi?id=267150
rdar://120557669

Reviewed by Tyler Wilcock.

Per spec changes (see https://github.com/w3c/aria/pull/2090), we should treat `aria-hidden=false` as undefined.

This patch removes support for aria-hidden=false, updating places where we relied on the behaviors of isNodeARIAVisible
with the new behavior. For example, we need to check if a child node is focused before deciding whether to skip it if it is
aria-hidden (tested by accessibility/datetime/input-date-field-labels-and-value-changes.html).

Tests that explicitly validate aria-hidden false were removed, and a new test to check that we are ignoring this property
has been added.

* LayoutTests/accessibility/aria-hidden-false-ignored-expected.txt: Added.
* LayoutTests/accessibility/aria-hidden-false-ignored.html: Added.
* LayoutTests/accessibility/aria-hidden-false-works-in-subtrees.html: Removed.
* LayoutTests/accessibility/aria-hidden-negates-no-visibility.html: Removed.
* LayoutTests/accessibility/aria-hidden-subtree-expected.txt: Added.
* LayoutTests/accessibility/aria-hidden-subtree.html: Added.
* LayoutTests/accessibility/aria-modal-expected.txt:
* LayoutTests/accessibility/aria-modal.html:
* LayoutTests/accessibility/aria-visible-element-roles.html: Removed.
* LayoutTests/accessibility/datetime/input-date-field-labels-and-value-changes.html:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::modalElementHasAccessibleContent):
(WebCore::AXObjectCache::isNodeVisible const):
(WebCore::AXObjectCache::getOrCreate):
(WebCore::isNodeFocused):
(WebCore::isNodeAriaVisible): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::textUnderElement const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::defaultObjectInclusion const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::addNodeOnlyChildren):

Canonical link: https://commits.webkit.org/284905@main
